### PR TITLE
Codespeak

### PR DESCRIPTION
**Pull request by SirRichardFrancis**
## About The Pull Request
Ironhammer got "codespeak" perk, that adds new tab to their interface, it contains verbs that "say" predefined phrase in security channel, most of those include user's location, and some used without radio.
Resulting message looks like some sort of police code, but only characters with "codespeak" would understand what it actually mean.

image
image
image
image
image

~Currently perk has placeholder description, and list of codes rather small, so I'm looking for ideas.~

image

## Why It's Good For The Game

Yet another feature.

## Changelog
:cl:
add: code language for IH
imageadd: Image for codespeak perk
/:cl:

### CODE REVIEW

**Comment by SaltyOldBull** [comment]
Making these codes randomized would go a long way to keep powergamers from cracking IronHammer encryption by just memorizing the codes.

**Comment by SirRichardFrancis** [comment]
> Making these codes randomized would go a long way to keep powergamers from cracking IronHammer encryption by just memorizing the codes.

Codes are random, no connection to verb used. While unrealistic, that indeed prevents code from being cracked by non-ironhammer.

**Comment by SaltyOldBull** [comment]
Oh, very nice

**Comment by thedarkcrusader** [comment]
yes YES YES YES YES YES YES YES YES YES YES YES

**Comment by pexel5** [comment]
DEAR LORD, YES, YES, YEEEEEEEEES!

**Comment by Paket1337** [comment]
Very cool but I dont see any code for contractor

**Comment by Alliostra** [comment]
I can already see operatives no longer talking to crewies at all, very nice

**Comment by Jan-Blasiak** [comment]
[insert excited response here]

**Review by Firefox13 — COMMENTED** [review]
Might want to make a delay on the codespeak, it's marginally easier to spam using auto-clicker rather than setting up a macros.
Also, it looks like this when I done local testing. Are you sure everything is right?
изображение

**Comment by SirRichardFrancis** [comment]
Eh? `" (" + replace_characters(message, list("¿"=")"))` should cut out that symbol, and it worked for me so far. Let's see...

**Comment by SirRichardFrancis** [comment]
Yeah, working on my side.
image

**Comment by Firefox13** [comment]
> try cloning "codespeak" branch from my fork.

Literally what I did. Whatever then, it's probably unicode problem

**Comment by SirRichardFrancis** [comment]
> Literally what I did. Whatever then, it's probably unicode problem

Yeah, seems like "¿" thing was lost for some reason. But I need some special symbol for this system to work, that is, if I don't want to rewrite "say" code, and I certainly do not.

**Review by Firefox13 — APPROVED** [review]

**Comment by SirRichardFrancis** [comment]
Codes displayed broken in telecom logs, so I did this. Could make it check user's perks and stuff, but that's too much trouble for a thing that never would be used.
image

File: code/modules/mob/language/codespeak.dm [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+/mob/living/carbon/human/proc/codespeak_no_local()
+	set category = "Codespeak"
+	set name = "(local) No"
+	src.codesay("Negative", FALSE, TRUE)
+
+/mob/living/carbon/human/proc/codespeak_warcrime_local()
+	set category = "Codespeak"
+	set name = "(local) Attack?"
+	src.codesay("We attack?", FALSE, TRUE) //TODO: Replace with something that sounds good
```
Comment (TheShown911) [inline code/modules/mob/language/codespeak.dm:135]: Maybe `Engage?` or `Engage target?`

**Review by TheShown911 — COMMENTED** [review]
There should be a way for new recruits to learn the codes or antags to steal the codes. Maybe a secret book of codes that's just and oddity with this perk hidden in the LT office or the doughnuts from the IH faction items letting you understand the codes for a while.

File: code/modules/mob/language/codespeak.dm (hunk 1 above)
Comment (SirRichardFrancis) [reply code/modules/mob/language/codespeak.dm]: Tried that, looks wanky too. But all better than "attack" I guess. Replaced.

**Comment by SirRichardFrancis** [comment]
> There should be a way for new recruits to learn the codes or antags to steal the codes. Maybe a secret book of codes that's just and oddity with this perk hidden in the LT office or the doughnuts from the IH faction items letting you understand the codes for a while.

Planning to add subtype of "metnal imprinter" that will do the thick to IH shop from GLF doc. When it's implemented, that is.
Maybe to antag uplink as well, for lots of telecrystals. But only after IH gets it.
Having recruits as IH is rare anyway, shouldn't be that much of an issue.

**Pull request merged**